File: src/models/proposed_transaction_edit.rs
use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Deserializer, Serialize};

use super::{Id, TransactionAnnotationPatch};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProposedTransactionEditStatus {
    Pending,
    Approved,
    Rejected,
    Removed,
}

/// A queued transaction annotation edit.
///
/// The editable fields have the same tri-state semantics as
/// `TransactionAnnotationPatch`: absent means no change, null means clear, and a
/// value means set.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(from = "ProposedTransactionEditSerde")]
pub struct ProposedTransactionEdit {
    pub id: Id,
    pub account_id: Id,
    pub transaction_id: Id,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub status: ProposedTransactionEditStatus,

    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_patch_field"
    )]
    pub description: Option<Option<String>>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_patch_field"
    )]
    pub note: Option<Option<String>>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_patch_field"
    )]
    pub tags: Option<Option<Vec<String>>>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_patch_field"
    )]
    pub subtags: Option<Option<Vec<String>>>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_patch_field"
    )]
    pub effective_date: Option<Option<NaiveDate>>,
}

#[derive(Debug, Deserialize)]
struct ProposedTransactionEditSerde {
    id: Id,
    account_id: Id,
    transaction_id: Id,
    created_at: DateTime<Utc>,
    updated_at: DateTime<Utc>,
    status: ProposedTransactionEditStatus,

    #[serde(default, deserialize_with = "deserialize_patch_field")]
    description: Option<Option<String>>,
    #[serde(default, deserialize_with = "deserialize_patch_field")]
    note: Option<Option<String>>,
    #[serde(default, deserialize_with = "deserialize_patch_string_or_vec_field")]
    category: Option<Option<Vec<String>>>,
    #[serde(default, deserialize_with = "deserialize_patch_string_or_vec_field")]
    subcategory: Option<Option<Vec<String>>>,
    #[serde(default, deserialize_with = "deserialize_patch_field")]
    tags: Option<Option<Vec<String>>>,
    #[serde(default, deserialize_with = "deserialize_patch_field")]
    subtags: Option<Option<Vec<String>>>,
    #[serde(default, deserialize_with = "deserialize_patch_field")]
    effective_date: Option<Option<NaiveDate>>,
}

impl From<ProposedTransactionEditSerde> for ProposedTransactionEdit {
    fn from(value: ProposedTransactionEditSerde) -> Self {
        Self {
            id: value.id,
            account_id: value.account_id,
            transaction_id: value.transaction_id,
            created_at: value.created_at,
            updated_at: value.updated_at,
            status: value.status,
            description: value.description,
            note: value.note,
            tags: value.tags.or(value.category),
            subtags: value.subtags.or(value.subcategory),
            effective_date: value.effective_date,
        }
    }
}

fn deserialize_patch_field<'de, D, T>(deserializer: D) -> Result<Option<Option<T>>, D::Error>
where
    D: Deserializer<'de>,
    T: Deserialize<'de>,
{
    Ok(Some(Option::<T>::deserialize(deserializer)?))
}

fn deserialize_patch_string_or_vec_field<'de, D>(
    deserializer: D,
) -> Result<Option<Option<Vec<String>>>, D::Error>
where
    D: Deserializer<'de>,
{
    #[derive(Deserialize)]
    #[serde(untagged)]
    enum StringOrVec {
        String(String),
        Vec(Vec<String>),
    }

    Ok(Some(
        match Option::<StringOrVec>::deserialize(deserializer)? {
            None => None,
            Some(StringOrVec::String(value)) => Some(vec![value]),
            Some(StringOrVec::Vec(values)) => Some(values),
        },
    ))
}

impl ProposedTransactionEdit {
    pub fn has_edit(&self) -> bool {
        self.description.is_some()
            || self.note.is_some()
            || self.tags.is_some()
            || self.subtags.is_some()
            || self.effective_date.is_some()
    }

    pub fn with_status(&self, status: ProposedTransactionEditStatus, now: DateTime<Utc>) -> Self {
        Self {
            status,
            updated_at: now,
            ..self.clone()
        }
    }

    pub fn to_annotation_patch(&self, timestamp: DateTime<Utc>) -> TransactionAnnotationPatch {
        TransactionAnnotationPatch {
            transaction_id: self.transaction_id.clone(),
            timestamp,
            description: self.description.clone(),
            note: self.note.clone(),
            tags: self.tags.clone(),
            subtags: self.subtags.clone(),
            effective_date: self.effective_date,
        }
    }
}

File: ts/src/models/proposed-transaction-edit.ts
import { Id } from './id.js';
import {
  type TransactionAnnotationPatchType,
  type TransactionAnnotationPatchJSON,
} from './transaction-annotation.js';

export type ProposedTransactionEditStatus = 'pending' | 'approved' | 'rejected' | 'removed';

export interface ProposedTransactionEditType {
  readonly id: Id;
  readonly account_id: Id;
  readonly transaction_id: Id;
  readonly created_at: Date;
  readonly created_at_raw?: string;
  readonly updated_at: Date;
  readonly updated_at_raw?: string;
  readonly status: ProposedTransactionEditStatus;
  readonly description?: string | null;
  readonly note?: string | null;
  readonly category?: string | null;
  readonly subcategory?: string | null;
  readonly tags?: string[] | null;
  readonly effective_date?: string | null;
}

export interface ProposedTransactionEditJSON {
  id: string;
  account_id: string;
  transaction_id: string;
  created_at: string;
  updated_at: string;
  status: ProposedTransactionEditStatus;
  description?: string | null;
  note?: string | null;
  category?: string | null;
  subcategory?: string | null;
  tags?: string[] | null;
  effective_date?: string | null;
}

export const ProposedTransactionEdit = {
  toJSON(edit: ProposedTransactionEditType): ProposedTransactionEditJSON {
    const json: ProposedTransactionEditJSON = {
      id: edit.id.toJSON(),
      account_id: edit.account_id.toJSON(),
      transaction_id: edit.transaction_id.toJSON(),
      created_at: edit.created_at_raw ?? edit.created_at.toISOString(),
      updated_at: edit.updated_at_raw ?? edit.updated_at.toISOString(),
      status: edit.status,
    };
    if (edit.description !== undefined) json.description = edit.description;
    if (edit.note !== undefined) json.note = edit.note;
    if (edit.category !== undefined) json.category = edit.category;
    if (edit.subcategory !== undefined) json.subcategory = edit.subcategory;
    if (edit.tags !== undefined) json.tags = edit.tags;
    if (edit.effective_date !== undefined) json.effective_date = edit.effective_date;
    return json;
  },

  fromJSON(json: ProposedTransactionEditJSON): ProposedTransactionEditType {
    return {
      id: Id.fromString(json.id),
      account_id: Id.fromString(json.account_id),
      transaction_id: Id.fromString(json.transaction_id),
      created_at: new Date(json.created_at),
      created_at_raw: json.created_at,
      updated_at: new Date(json.updated_at),
      updated_at_raw: json.updated_at,
      status: json.status,
      description: json.description,
      note: json.note,
      category: json.category,
      subcategory: json.subcategory,
      tags: json.tags,
      effective_date: json.effective_date,
    };
  },

  withStatus(
    edit: ProposedTransactionEditType,
    status: ProposedTransactionEditStatus,
    now: Date,
  ): ProposedTransactionEditType {
    return {
      ...edit,
      status,
      updated_at: now,
      updated_at_raw: undefined,
    };
  },

  toAnnotationPatch(
    edit: ProposedTransactionEditType,
    timestamp: Date,
  ): TransactionAnnotationPatchType {
    return {
      transaction_id: edit.transaction_id,
      timestamp,
      ...(edit.description !== undefined ? { description: edit.description } : {}),
      ...(edit.note !== undefined ? { note: edit.note } : {}),
      ...(edit.category !== undefined ? { category: edit.category } : {}),
      ...(edit.subcategory !== undefined ? { subcategory: edit.subcategory } : {}),
      ...(edit.tags !== undefined ? { tags: edit.tags } : {}),
      ...(edit.effective_date !== undefined ? { effective_date: edit.effective_date } : {}),
    };
  },

  patchJSON(
    edit: ProposedTransactionEditType,
  ): Omit<TransactionAnnotationPatchJSON, 'transaction_id' | 'timestamp'> {
    return {
      ...(edit.description !== undefined ? { description: edit.description } : {}),
      ...(edit.note !== undefined ? { note: edit.note } : {}),
      ...(edit.category !== undefined ? { category: edit.category } : {}),
      ...(edit.subcategory !== undefined ? { subcategory: edit.subcategory } : {}),
      ...(edit.tags !== undefined ? { tags: edit.tags } : {}),
      ...(edit.effective_date !== undefined ? { effective_date: edit.effective_date } : {}),
    };
  },
} as const;
